### ISSUE TITLE
Formik 2 compatibility

### ISSUE THREAD

**Issue by Kuzirashi** [stackworx/formik-material-ui#111]
After Formik 2 has been released, is this package compatible?

**Comment by cliedeman** [comment]
No I don’t think so. After a quick look at the new docs Field is not used? My plan is to release current as 1.0 and then 2.0 compatible with formik 2

**Comment by maxcan** [comment]
following.. would love this

**Comment by amoe** [comment]
Why do we think that `Field` is not used?  From a brief look at the [migration guide](https://github.com/jaredpalmer/formik/releases/tag/v2.0.1) it seems that `<Field>` is still present in a lot of the examples.

**Comment by cliedeman** [comment]
It not that you can't use it, its that the new useField style does not need it. It's the same as putting a `<Field.../>` inside the wrapped component. 

See https://jaredpalmer.com/formik/docs/api/useField#docsNav.

```
<form onSubmit={props.handleSubmit}>
          <MyTextField name="firstName" type="text" label="First Name" />
          <MyTextField name="lastName" type="text" label="Last Name" />
          <MyTextField name="email" type="email" label="Email" />
          <button type="submit">Submit</button>
        </form>```

The nice thing about this style is that you get better type safety. Granted you could do this in formik 1

**Comment by Favna** [comment]
I managed to make everything work in our implementation. The basic gist is that instead of using the component or render props of a field we instead pass the formik-material-ui based component as children to a `<FastField>`.

```tsx
// Child of Form, which itself is child of Formik. Both Formik components.

<FastField
  name='email' type='email' label='email'
  onChange={handleChange} onBlur={handleBlur}
  className={errors.email && touched.email ? classnames(css.input, 'error') : 'text-input'}
>
  {fieldProps => <ValidatedFormTextField {...fieldProps} variant='outlined' required />}
</FastField>
```

`<ValidatedFormTextField>` is as follows:

```tsx
import { TextField } from '@material-ui/core';
import { objectHasProperty } from '@rws-air/utils';
import { fieldToTextField, TextFieldProps } from 'formik-material-ui';
import React, { FC } from 'react';
import css from 'ValidatedFormTextField.module.scss';

const ValidatedFormTextField: FC<TextFieldProps> = props => {
  return (
    <TextField
      {...fieldToTextField(props)}
      value={props.field.value}
      label={props.label}
      placeholder={props.placeholder}
      onChange={props.field.onChange}
      onBlur={props.field.onBlur}
      margin='dense'
      variant='outlined'
      className={props.className}
      autoFocus={objectHasProperty(props, 'autoFocus') ? props.autoFocus : false}
      fullWidth
      required={objectHasProperty(props, 'required') || false}
      InputProps={{
        classes: {
          root: css.input,
          notchedOutline: css.outline,
        },
      }}
      InputLabelProps={{
        shrink: true,
        classes: { root: css.inputLabel },
      }}
      FormHelperTextProps={
        {
          component: 'div',
          classes:
              { error: css.errorLabel },
        }
      }
    />
  );
};
```

**Comment by DjolePetrovic** [comment]
Any updates on this topic? Do you guys need some help?

**Comment by cliedeman** [comment]
Please try 2.0.0-alpha.2

**Comment by cliedeman** [comment]
2.0.0-alpha.2 is formik 2 compatible. Please give it a try

**Comment by Favna** [comment]
I'll give it a thorough try somewhere this week, thought right away I can say that after a quick upgrade I got a TS error saying that "fieldToSelect" was not exported from formik-material-ui. I didn't have time to look into things further just yet.

**Comment by cliedeman** [comment]
Migration notes are [here](https://stackworx.github.io/formik-material-ui/docs/guide/migrating)

**Comment by Favna** [comment]
ah got it, yeah we use customized components so I'll have to look into it more thoroughly then, can't really give any timebox. Thanks for the link.

**Comment by Favna** [comment]
@ciaran-stackworx I've tried to upgrade out custom components but without success. I've pushed my changes to a branch, one of the components can be found here: https://github.com/RWS-NL/air-node-packages/blob/chore/upgrade-formik/packages/webcomponents/src/FormikComponents/RadioGroup/index.tsx. The others are 1 folder back (to the "FormikComponents" folder) and in each of the subfolders there is a custom component. If you could take a look at point me in the direction that I'm supposed to go here I would really appreciate it. As example of implementation, [the TextField component](https://github.com/RWS-NL/air-node-packages/blob/chore/upgrade-formik/packages/webcomponents/src/FormikComponents/TextField/index.tsx). In the final code I'd imagine it would be used like so, which at least is our current implementation and afaik still the  recommended way for Formik:

```tsx
<Formik {...formikConfig}>
      {formikProps => {
        const { handleSubmit } = formikProps;

        return (
        <Form name='sample-form'>
          <FastField
            name='some-name'
            type='text'
            label='some-label'
            placeholder='some-placeholder'
            variant='outlined'
            data-qa='formFieldSomeName' // We use data-qa for testing
            component={TextField}
            required />
        </Form>
// And a lot more here
      )
}}
</Formik>
```

**Comment by cliedeman** [comment]
@Favna I followed the example under [React Context](https://jaredpalmer.com/formik/docs/tutorial#leveraging-react-context)

Snippet from the docs lower down

> React is all about composition, and while we've cut down on a lot of the prop-drilling, we are still repeating ourselves with a label, <Field>, and <ErrorMessage> for each of our inputs. We can do better with an abstraction! With Formik, you can and should build reusable input primitive components that you can share around your application. Turns out our <Field> render-prop component has a sister and her name is useField that's going to do the same thing, but via React Hooks! Check this out...

Can you change your example to this

```
<Form name='sample-form'>
          <TextField
            name='some-name'
            type='text'
            label='some-label'
            placeholder='some-placeholder'
            variant='outlined'
            data-qa='formFieldSomeName' // We use data-qa for testing
              required />
        </Form>
```

**Comment by Favna** [comment]
@cliedeman Thank you for that suggestion. That appears to work perfectly. I've got just got some troubles left with the DatePicker now, which uses formik-material-ui-pickers now.

TypeScript is telling me that I cannot pass an `onChange` nor `error` property to the `KeyboardDatePicker` from the library. I feel like this is probably a mistake. I believe this is in part the reason that when entering an invalid date in the date picker with keyboard input, the FormHelperText does show up, but it's not red until focus is lost from the KeyboardDatePicker: 
![image](https://user-images.githubusercontent.com/4019718/72817036-67dd2400-3c69-11ea-8703-e6037bdf9764.png)

This is our DatePicker component: [link here](https://github.com/RWS-NL/air-node-packages/blob/chore/upgrade-formik/packages/webcomponents/src/FormikComponents/DatePicker/index.tsx)

And this is my current implementation (of course inside a Form and Formik component like before): 
```tsx
<DatePicker minDate={new Date('01-01-2012')} name='date' type='text' label='Sample Date' />
```

Also you'll see that despite `new Date('01-01-2012')` resolving to 1st of January 2012 officially, the component shows `20-01-2020` (I removed the last 0 in the screenshot)
```js
new Date('01-01-2012') // ==> Sun Jan 01 2012 00:00:00 GMT+0100 (Central European Standard Time)
```

**Comment by cliedeman** [comment]
@Favna thanks for kicking the tyres on this.

The date picker link is broken?

I have also been reading the Formik source code and think the custom text field approach above is not correct. [useField](https://github.com/jaredpalmer/formik/blob/master/packages/formik/src/Field.tsx#L97) should not be invoked twice.

I have [refactored](https://github.com/stackworx/formik-material-ui/commit/72c3b6ef1698f92c046619486c0d4009ee92de49) the code and updated the [custom component documentation](https://stackworx.github.io/formik-material-ui/docs/guide/custom-component/)

New Releases:
 - formik-material-ui-pickers@0.0.4
 - formik-material-ui@2.0.0-alpha.3

**Comment by Favna** [comment]
@cliedeman I just fixed the link to the datepicker component. I had since merged the branch back to master which removed the branch. I'll have to take a look into the new release later, quite busy the coming days though I'm afraid.

**Comment by cliedeman** [comment]
No rush, I will keep the release in alpha/beta while I test out the API and make sure I am happy with it

**Comment by Favna** [comment]
@cliedeman I just checked out the latest releases. Reporting my findings here

1. I can now correctly pass the `onChange` property to `KeyboardDatePicker`. It works perfectly now.
  - The issue with the FormHelperText not properly turning red was a derp on my side in that I forgot the CSS style color, woops
  - The issue where I thought minDate wasn't being set correctly was also a derp, still adjusting to 2020 instead of 2019 lol.
  - I actually don't even need the onChange properly to be honest, but it's good that it's added for others.

2. As for not calling `useField` twice, there isn't really any way around that in TypeScript without completely hacking up the proptypes to convince TS that the Formik specific props are available. For example in our TextField component, the props are of type `TextFieldProps` (from formik-material-ui) but we also need the `value` among other props which are not available on its type. Here's an example image but for a proper full-code example you can best check out the components I linked before, or follow this link to go to the folder with all the latest versions: https://github.com/RWS-NL/air-node-packages/tree/master/packages/webcomponents/src/FormikComponents

![image](https://user-images.githubusercontent.com/4019718/73072434-5720ee00-3eb5-11ea-8a89-5243c958d7a5.png)

**Comment by dickiedyce** [comment]
Hi, I was just wondering if the migration instructions were still correct? Specifically for custom fields?

![formik-material-ui-migrations](https://user-images.githubusercontent.com/1810481/73640401-1455d780-4666-11ea-9179-0213a33253e1.png)

**Comment by cliedeman** [comment]
There is an updated example [here](https://stackworx.github.io/formik-material-ui/docs/guide/custom-component#upper-casing-field). This [one is outdated](https://stackworx.github.io/formik-material-ui/docs/guide/migrating#after-1)

```
import MuiTextField from '@material-ui/core/TextField';
import {
  useFieldToTextField,
  TextField,
  TextFieldProps,
} from 'formik-material-ui';
const UpperCasingTextField = (props: TextFieldProps) => {
  const customize = React.useCallback(([, , helpers]) => {
    return {
      onChange: (event: React.ChangeEvent<HTMLInputElement>) => {
        const { value } = event.target;
        helpers.setValue(value ? value.toUpperCase() : '');
      },
    };
  }, []);
  return <MuiTextField {...useFieldToTextField(props, customize)} />;
};
```

**Comment by Favna** [comment]
@dickiedyce adding to Cieran's reply you can also take a look at how we have implemented it, as example, Text field: https://github.com/RWS-NL/air-node-packages/blob/master/packages/webcomponents/src/FormikComponents/TextField/index.tsx

The code also has TSDoc comments showing how it is used.

**Comment by cliedeman** [comment]
After giving this some thought I have decided to revert the useField changes and go back the the 1.0 API style. 

The changes have been published as 2.0.0-beta.0 (0.0.5 for pickers).

The 2 main reasons were:

- `useField` and `FastField` cannot be combined
- useField has no way to pass in properties like type and multiple. These properties are important in Formik 2 and change how internal state is managed.

Please see the updated (and now much shorter...) migration [docs](https://stackworx.github.io/formik-material-ui/docs/guide/migrating)

Sorry for all the churn and thanks to everyone who helped test.

Ciaran

**Comment by cliedeman** [comment]
I am going to close this as Formik 2 now works as expected.

The last decision before I release the final version is wether or not I wrap all the components with Forward ref

References

- https://reactjs.org/docs/forwarding-refs.html
- https://www.newline.co/fullstack-react/articles/using-refs-in-react/#forwarding-refs
